src/main/java/com/aws/greengrass/authorization/WildcardVariableTrie.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.authorization;
+
+import com.aws.greengrass.util.DefaultConcurrentHashMap;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Wildcard trie node which contains properties to identify the Node and a map of all it's children.
+ * - isTerminal: If the node is a terminal node while adding a resource. It might not necessarily be a leaf node as we
+ *   are adding multiple resources having same prefix but terminating on different points.
+ * - isTerminalLevel: If the node is the last level before a valid use "#" wildcard (eg: "abc/123/#", 123/ would be the
+ *   terminalLevel).
+ * - isWildcard: If current Node is glob wildcard (*)
+ * - isWildcard: If current Node is MQTT wildcard (#, +)
+ * - isNull: If null
+ * - matchAll: if current node should match everything. Could be MQTTWildcard or a wildcard and will always be a
+ *   terminal Node.
+ */
+public class WildcardVariableTrie {
+    protected static final char GLOB_WILDCARD = '*';
+    protected static final char MQTT_MULTILEVEL_WILDCARD = '#';
+    protected static final char MQTT_SINGLELEVEL_WILDCARD = '+';
+    protected static final char MQTT_LEVEL_SEPARATOR = '/';
+
+    private boolean isTerminal;
+    private boolean isTerminalLevel;
+    private boolean isWildcard;
+    private boolean isMQTTWildcard;
+    private boolean isNull;
+    private boolean matchAll;
+    private final Map<String, WildcardVariableTrie> children =
+            new DefaultConcurrentHashMap<>(WildcardVariableTrie::new);
+
+    /**
+     * Add allowed resources for a particular operation.
+     * - A new node is created for every occurrence of a wildcard (*, #, +).
+     * - Only nodes with valid usage of wildcards are marked with isWildcard or isMQTTWildcard.
+     * - Any other characters are grouped together to form a node.
+     * - Just a '*' or '#' creates a Node setting matchAll to true and would match all resources
+     *
+     * @param subject resource pattern
+     */
+    public void add(String subject) {
+        if (subject == null) {
+            isNull = true;
+            return;
+        }
+        if (subject.equals(String.valueOf(GLOB_WILDCARD))) {
+            WildcardVariableTrie initial = this.children.get(String.valueOf(GLOB_WILDCARD));
+            initial.matchAll = true;
+            initial.isTerminal = true;
```

Comment:
glob wildcard isn't terminal now is it? Since it should stop matching at `/`

Reply:
Just Glob should be a match everything, as we previously did that.

Reply:
This is testing for `*` and not `Xyz*`?
Since this is recursive, how are you handling it in the case of `a/*/c`? `a/b/c` is good, but `a/b/d/e` shouldn't match

Reply:
Also in normal cases while matching terminal "*" we do check if it doesn't match a "/"

Reply:
But if it is standalone, `*` then it must match `/` right?

Reply:
> This is testing for `*` and not `Xyz*`? Since this is recursive, how are you handling it in the case of `a/*/c`? `a/b/c` is good, but `a/b/d/e` shouldn't match

yes thats correct, first Add is not recursive, only the second one is. So only when its just `*` we make it a matchAll, wildCard and terminal. In all other cases we just make it a wildCard or terminal (if terminal)